Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool.swift

```diff
@@ -121,6 +128,364 @@ enum HTTPConnectionPool {
+            connection.executeRequest(request.req)
+            if cancelTimeout {
+                self.cancelRequestTimeout(request.id)
+            }
+
+        case .executeRequestsAndCancelTimeouts(let requests, let connection):
+            for request in requests {
+                connection.executeRequest(request.req)
+                self.cancelRequestTimeout(request.id)
+            }
+
+        case .failRequest(let request, let error, cancelTimeout: let cancelTimeout):
+            if cancelTimeout {
+                self.cancelRequestTimeout(request.id)
+            }
+            request.req.fail(error)
+
+        case .failRequestsAndCancelTimeouts(let requests, let error):
+            for request in requests {
+                self.cancelRequestTimeout(request.id)
+                request.req.fail(error)
+            }
+
+        case .scheduleRequestTimeout(let request, on: let eventLoop):
+            self.scheduleRequestTimeout(request, on: eventLoop)
+
+        case .cancelRequestTimeout(let requestID):
+            self.cancelRequestTimeout(requestID)
+
+        case .none:
+            break
+        }
+    }
+
+    // MARK: Run actions
+
+    private func createConnection(_ connectionID: Connection.ID, on eventLoop: EventLoop) {
+        self.connectionFactory.makeConnection(
+            for: self,
+            connectionID: connectionID,
+            http1ConnectionDelegate: self,
+            http2ConnectionDelegate: self,
+            deadline: .now() + (self.clientConfiguration.timeout.connect ?? .seconds(30)),
+            eventLoop: eventLoop,
+            logger: self.logger
+        )
+    }
+
+    private func scheduleRequestTimeout(_ request: Request, on eventLoop: EventLoop) {
+        let requestID = request.id
+        let scheduled = eventLoop.scheduleTask(deadline: request.connectionDeadline) {
+            // The timer has fired. Now we need to do a couple of things:
+            //
+            // 1. Remove ourselves from the timer dictionary to not leak any data. If our
+            //    waiter entry still exist, we need to tell the state machine, that we want
+            //    to fail the request.
+
+            let timeout = self.timerLock.withLock {
+                self._requestTimer.removeValue(forKey: requestID) != nil
+            }
+
+            // 2. If the entry did not exists anymore, we can assume that the request was
+            //    scheduled on another connection. The timer still fired anyhow because of a
+            //    race. In such a situation we don't need to do anything.
+            guard timeout else { return }
+
+            // 3. Tell the state machine about the timeout
+            let action = self.stateLock.withLock {
+                self._state.timeoutRequest(requestID)
+            }
+
+            self.run(action: action)
+        }
+
+        self.timerLock.withLockVoid {
+            assert(self._requestTimer[requestID] == nil)
```

Comment:
Doesn't matter much because it's only an assert, but you can also check the return value from `updateValue(scheduled, forKey: requestID)` is `nil`